util/src/main/java/io/kubernetes/client/util/KubeConfig.java

```diff
@@ -229,6 +244,99 @@ public String getAccessToken() {
     return null;
   }
 
+  /**
+   * Attempt to create an access token by running a configured external program.
+   *
+   * @see <a
+   *     href="https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins">
+   *     Authenticating » client-go credential plugins</a>
+   */
+  @SuppressWarnings("unchecked")
+  private String tokenViaExecCredential(Map<String, Object> execMap) {
+    if (execMap == null) {
+      return null;
+    }
+    String apiVersion = (String) execMap.get("apiVersion");
+    if (!"client.authentication.k8s.io/v1beta1".equals(apiVersion)
+        && !"client.authentication.k8s.io/v1alpha1".equals(apiVersion)) {
+      log.error("Unrecognized user.exec.apiVersion: {}", apiVersion);
+      return null;
+    }
+    String command = (String) execMap.get("command");
+    JsonElement root = runExec(command, (List) execMap.get("args"), (List) execMap.get("env"));
+    if (root == null) {
+      return null;
+    }
+    if (!"ExecCredential".equals(root.getAsJsonObject().get("kind").getAsString())) {
+      log.error("Unrecognized kind in response");
+      return null;
+    }
+    if (!apiVersion.equals(root.getAsJsonObject().get("apiVersion").getAsString())) {
+      log.error("Mismatched apiVersion in response");
+      return null;
+    }
+    JsonObject status = root.getAsJsonObject().get("status").getAsJsonObject();
+    JsonElement token = status.get("token");
+    if (token == null) {
+      // TODO handle clientCertificateData/clientKeyData
+      // (KubeconfigAuthentication is not yet set up for that to be dynamic)
+      log.warn("No token produced by {}", command);
+      return null;
+    }
+    log.debug("Obtained a token from {}", command);
+    return token.getAsString();
+    // TODO cache tokens between calls, up to .status.expirationTimestamp
+    // TODO a 401 is supposed to force a refresh,
+    // but KubeconfigAuthentication hardcodes AccessTokenAuthentication which does not support that
+    // and anyway ClientBuilder only calls Authenticator.provide once per ApiClient;
+    // we would need to do it on every request
+  }
+
+  private JsonElement runExec(String command, List<String> args, List<Map<String, String>> env) {
+    List<String> argv = new ArrayList<>();
+    if (command.startsWith("./")) {
```

Comment:
This is Linux specific, use `File.isAbsolute(command)` instead.

Reply:
and also, presumably you want to catch `../foo/bar` too which this doesn't currently cover.

Reply:
> This is Linux specific

Not really. It should work fine on Windows as well.

> use `File.isAbsolute(command)` instead

That is not going to work since the _normal_ case is a plain command from the path like `aws-iam-authenticator`.

> presumably you want to catch `../foo/bar` too

As noted in the TODO comment, it is unclear. The [official documentation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuration) does not give any definition of “[r]elative command paths”. Is `bin/script` to be treated this way, or only `./bin/script`? `.\bin\script`? Is `/usr/local/bin/script` legal, despite clearly not being “relative”? Test cases for the Golang client which actually run a script [only cover](https://github.com/kubernetes/client-go/blob/9c9f7f424e65af6ded246bd4ba1a1cec988acc7b/plugin/pkg/client/auth/exec/exec_test.go#L639) the `./` prefix. From the [Golang implementation](https://github.com/kubernetes/client-go/blob/9c9f7f424e65af6ded246bd4ba1a1cec988acc7b/plugin/pkg/client/auth/exec/exec.go#L304-L312) it is not even clear that this system works at all: it appears to resolve paths against the process current working directory, which might be completely unrelated to the location of `$KUBECONFIG`—I presume the test passes merely because the Go test runner changes directory to the parent of `exec_test.go`.

Reply:
Experimentally, `kubectl` seems to work with for example `../bin/gke-auth` even when run from an arbitrary directory. Unclear to me from looking at sources how that works. At any rate, the best I can think of is to treat any `command` containing either `/` or `\` as a (possibly absolute) file path to be resolved against the `$(basedir $KUBECONFIG)`.